### ISSUE TITLE
Application code needing both Netconf Client and Server functionality

### ISSUE THREAD

**Issue by ankit-ka** [sysrepo/sysrepo#1790]
Hi,

I have an application (A1) that runs on a host machine (M1) and this application needs to use Netconf interfaces to two other applications running on two different machines (say Application A2 which runs on machine M2 and behaves like Netconf client for A1 and applicaiton A3 which runs on machine M3 and behaves like Netconf server for A1). 

A2 (Netconf Client) <======> A1 (Netconf Client and Server) <=======> A3 (Netconf Server) 

Now A1 needs to behave like a Netconf Server to A2 and exchange data with it. And the same A1 needs to behave like a Netconf client to A3 and exchange data with it. 

So the questions are:

1. Will the sysrepo library on machine M1 manage different set of datastores for the Netconf server and Netconf client functionalities? If yes, how are these different datastores identified?

2. If this application A1 code has to send an rpc message to Netconf Client instance on A2 or to Netconf Server instance on A3, how will A1 code differentiate where to send this rpc message? How does the application A1 invoke sysrepo APIs (say sr_rpc_send) to indicate where is the rpc-send destined to?

Please suggest any documents or example application code which illustrates such dual Netconf connectivity.

Thanks!
Ankit

**Comment by michalvasko** [comment]
Hi Ankit,
what is important here is to strictly distinguish between *sysrepo* and *netopeer2*. *sysrepo* itself is just a YANG datastore. You can imagine it as a complex database with some additional features. In this context, it is strictly a local machine software. So, if an application calls any `sr_*` function, it must be directly linked with *sysrepo* and is communicating with the local instance on the system, nothing more.

As for *netopeer2*, it is actually now a really simple application that just connects all the libraries to create a full NETCONF server stack. But all it does is translate remote NETCONF commands (functionality of *libnetconf2*) to local *sysrepo* commands.

But to answer you
1. NETCONF client does not manage its own datastore but rather remotely the datastore of the NETCONF server it connects to. So, if you run *netopeer2-cli* on machine `M1` and connect to the NETCONF server on localhost, you will effectively manage the datastore (*sysrepo*) on `M1`. If you connect remotely to `M3`, you will manage datastore of `M3`.

2. Application `A1` is unable to send an RPC to `A2` because `A2` is a NETCONF client, you can only send RPCs from a NETCONF client to a NETCONF server. Where an RPC is sent is determined by the NETCONF session, so you can have a NETCONF client connected to 2 different NETCONF servers (*netopeer2-cli* does not allow this) and then communicate with both of them. If `A1` invokes `sr_rpc_send()`, it will not be any NETCONF operation but a *sysrepo* call operating on the local *sysrepo* instance.

The document I would suggest most is [NETCONF specification](https://tools.ietf.org/html/rfc6241) because it seems you have misunderstood how NETCONF protocol operates.

Regards,
Michal

**Comment by ankit-ka** [comment]
Michal,

Thanks for your reply. I guess I need to put forth my use-case more clearly. Please see some comments inline below:

> 
> 
> Hi Ankit,
> what is important here is to strictly distinguish between _sysrepo_ and _netopeer2_. _sysrepo_ itself is just a YANG datastore. You can imagine it as a complex database with some additional features. In this context, it is strictly a local machine software. So, if an application calls any `sr_*` function, it must be directly linked with _sysrepo_ and is communicating with the local instance on the system, nothing more.
> 


ANKIT: Ok, I understand that sysrepo manages local datastore and provides a set of sr_* APIs that an application code can invoke to access the local datastore maintained by sysrepo.


> As for _netopeer2_, it is actually now a really simple application that just connects all the libraries to create a full NETCONF server stack. But all it does is translate remote NETCONF commands (functionality of _libnetconf2_) to local _sysrepo_ commands.
> 
> But to answer you
> 
>     1. NETCONF client does not manage its own datastore but rather remotely the datastore of the NETCONF server it connects to. So, if you run _netopeer2-cli_ on machine `M1` and connect to the NETCONF server on localhost, you will effectively manage the datastore (_sysrepo_) on `M1`. If you connect remotely to `M3`, you will manage datastore of `M3`.
> 
>     2. Application `A1` is unable to send an RPC to `A2` because `A2` is a NETCONF client, you can only send RPCs from a NETCONF client to a NETCONF server. Where an RPC is sent is determined by the NETCONF session, so you can have a NETCONF client connected to 2 different NETCONF servers (_netopeer2-cli_ does not allow this) and then communicate with both of them. If `A1` invokes `sr_rpc_send()`, it will not be any NETCONF operation but a _sysrepo_ call operating on the local _sysrepo_ instance.
> 

ANKIT: Assume that my application A1 (on host M1) talks to another application A2 (on host M2) such that A2 is a Netconf Client and sends database updates to A1 on Netconf interface. I understand that the Netconf Server on A1's host M1 will receive these messages and internally call sysrepo APIs to store the updates in the local data structures on  M1. And if applicaiton A1 has created a sysrepo session S1 and a connection on the datastore and have subscribed for a callback for module change (sr_module_change_subscribe). then A1 application will know of such datastore updates and can invoke appropriate sr_* APIs to read the updated value from local data store. So, in summary, sysrepo provides APIs for applications to read updated attributes received from Netconf client to server. 

Now, if my same application A1 has a talk to another application A3 (on host M3) using Netconf interface then how to achieve this using sysrepo APIs? Note that here A1 behaves like a Netconf client towards A3. The questions are:

1)  You mentioned - "Where an RPC is sent is determined by the NETCONF session" - so does sysrepo provide mechansim to allow an application to call APIs to initiate rpc messages to A3's Netconf Server and receive replies? If yes, will this need a separate sysrepo session S2 and how to create that?
 
2) If point 1 above is a valid scenario, then when the replies come from the Netconf server on A3, I understand that the sysrepo on A1 will update the datastore on A1's host. How will the application A1 know of this reply and get to read the content received in the rpc-reply? can it do sr_module_change_subscribe with session S2 for the same?

> 
> The document I would suggest most is [NETCONF specification](https://tools.ietf.org/html/rfc6241) because it seems you have misunderstood how NETCONF protocol operates.
> 
> Regards,
> Michal

**Comment by michalvasko** [comment]
Hi Ankit,
> So, in summary, sysrepo provides APIs for applications to read updated attributes received from Netconf client to server.

Yes, that would be a correct summary.

> 1. You mentioned - "Where an RPC is sent is determined by the NETCONF session" - so does sysrepo provide mechansim to allow an application to call APIs to initiate rpc messages to A3's Netconf Server and receive replies? If yes, will this need a separate sysrepo session S2 and how to create that?

You are still mixing everything together. *sysrepo* will never provide any mechanism to initiate RPCs to anything else but the locally subscribed applications that called `sr_rpc_subscribe()`.

> 2. If point 1 above is a valid scenario, then when the replies come from the Netconf server on A3, I understand that the sysrepo on A1 will update the datastore on A1's host. How will the application A1 know of this reply and get to read the content received in the rpc-reply? can it do sr_module_change_subscribe with session S2 for the same?

I suppose the scenario is not valid the way you described it. Also, if a reply comes from a NETCONF server, the sender should not update its datastore because it functions only as the NETCONF client, which only sends RPCs and receives replies, nothing else.

This is my suggestion, please ignore *sysrepo* altogether for now. Try to explain your scenario only using the NETCONF terminology (what is the client, what the server). Because as soon as you want any remote communication, *sysrepo* on its own is unable to help you at all.

Regards,
Michal

**Comment by ankit-ka** [comment]
> 
> 
> Hi Ankit,
> 
> > So, in summary, sysrepo provides APIs for applications to read updated attributes received from Netconf client to server.
> 
> Yes, that would be a correct summary.
> 
> > 1. You mentioned - "Where an RPC is sent is determined by the NETCONF session" - so does sysrepo provide mechansim to allow an application to call APIs to initiate rpc messages to A3's Netconf Server and receive replies? If yes, will this need a separate sysrepo session S2 and how to create that?
> 
> You are still mixing everything together. _sysrepo_ will never provide any mechanism to initiate RPCs to anything else but the locally subscribed applications that called `sr_rpc_subscribe()`.
> 
> > 1. If point 1 above is a valid scenario, then when the replies come from the Netconf server on A3, I understand that the sysrepo on A1 will update the datastore on A1's host. How will the application A1 know of this reply and get to read the content received in the rpc-reply? can it do sr_module_change_subscribe with session S2 for the same?
> 
> I suppose the scenario is not valid the way you described it. Also, if a reply comes from a NETCONF server, the sender should not update its datastore because it functions only as the NETCONF client, which only sends RPCs and receives replies, nothing else.
> 
> This is my suggestion, please ignore _sysrepo_ altogether for now. Try to explain your scenario only using the NETCONF terminology (what is the client, what the server). Because as soon as you want any remote communication, _sysrepo_ on its own is unable to help you at all.

ANKIT: Yeah, I guess I will try and explain the scenario from what I want to achieve using Netconf protocol. And you may help me understand how I can do it using sysrepo and Netopeer.

I have an application A1 (on machine M1) which needs to talk Netconf to two other applications - A2 (on M2) and A3 (on M3). A1 gets data updates (edit) from A2 and also needs to send its existing data values to A2. Similaryly A1 need to read data on A3 and update data on A3. 

A2 (Netconf Client) <==> A1 (Netconf Server and Client) <==> A3 (Netconf Server)

Below are high level functional requirements for A1:

1. A1's Netconf server should be able to accept rpc get-config and edit-config messages from A2's netconf client and reply back. 
2. application A1 should be able to know when A1's Netconf server receives data updates from A2 Netconf client so that the application A1 can use the updated data values.
3. application A1 (using its Netconf Client functionality) should be able to connect with the A3's Netconf server and be able to send rpc get-config and edit-config to A3's Netconf Server. And then A1 application should be able to read the rpc replies from A3 so that it can use the data received in these reply messages.

I hope I could explain the scenario with these high level requirements. Please help me in understanding how I can use the sysrepo's APIs along with netopeer2 binaries for each of the three points above.

Thanks again for your time and help.

Regards
Ankit

> Regards,
> Michal

**Comment by michalvasko** [comment]
Hi Ankit,
okay, I think one last bit of information is needed. Do you want data (*sysrepo*) of `A1` (`M1`) to be in synchronization with `A3` (`M3`)? In other words, should data on `M1` be always the same as data on `M3`?

Regards,
Michal

**Comment by ankit-ka** [comment]
> 
> 
> Hi Ankit,
> okay, I think one last bit of information is needed. Do you want data (_sysrepo_) of `A1` (`M1`) to be in synchronization with `A3` (`M3`)? In other words, should data on `M1` be always the same as data on `M3`?

ANKIT: Yes, I will need the A1 application to get data updates from A3/M3.

> 
> Regards,
> Michal

**Comment by michalvasko** [comment]
Hi Ankit,
in that case this it would basically mean that every `edit-config` received by `A1` needs to also be sent to `A3` and vice versa. Specifically, creating a NETCONF connection to the other server in the `edit-config` callback. The rest is following.

1. Standard *netopeer2-server* with *sysrepo* functionality provides this.
2. For this `A1` simply needs to subscribe to the data of interest in *sysrepo*.
3. Standard NETCONF client functionality and you should be able to use *libnetconf2* to implement it. However, it is not clear whether this `edit-config`, for example, should then also be applied on `M1`, only on `M3`, or you tried to implement the synchronization of `M1` and `M3` this way and the edit was already applied on `M1`. For the last case the safest and easiest way of making sure `M1` and `M3` are synchronized is my initial suggestion.

To summarize, your scenario should be possible but is not a common one so you may encounter some unexpected problems.

Regards,
Michal

**Comment by ankit-ka** [comment]
> 
> 
> Hi Ankit,
> in that case this it would basically mean that every `edit-config` received by `A1` needs to also be sent to `A3` and vice versa. Specifically, creating a NETCONF connection to the other server in the `edit-config` callback. The rest is following.
> 
>     1. Standard _netopeer2-server_ with _sysrepo_ functionality provides this.
> 
>     2. For this `A1` simply needs to subscribe to the data of interest in _sysrepo_.
> 

ANKIT: Yes. And I do have the standard Netopeer2-server running and my applicaiton code (A1) is integrated with the sysrepo library to make the sr_* calls and its working fine.

>     3. Standard NETCONF client functionality and you should be able to use _libnetconf2_ to implement it. However, it is not clear whether this `edit-config`, for example, should then also be applied on `M1`, only on `M3`, or you tried to implement the synchronization of `M1` and `M3` this way and the edit was already applied on `M1`. For the last case the safest and easiest way of making sure `M1` and `M3` are synchronized is my initial suggestion.

ANKIT: I read up some more information on the web regarding sysrepo and Netopeer2 and after going through your responses to my queries, I guess now I understand this better. 

My application code cannot simply use any sysrepo APIs (or some APIs provided by netopeer-cli) to behave as a  Netconf client. For that my application code will have to use ssh (libssh), yang (libyang) and netconf functionality (libnetconf) directly within the code to implement the Netconf client functionality and attempt to connect to a Netconf server running on M3.

Thanks again for your patience in going through my queries and helping me understand this better.

Regards,
Ankit

> 
> 
> To summarize, your scenario should be possible but is not a common one so you may encounter some unexpected problems.
> 
> Regards,
> Michal

**Comment by michalvasko** [comment]
Hi Ankit,
> My application code cannot simply use any sysrepo APIs (or some APIs provided by netopeer-cli) to behave as a Netconf client. For that my application code will have to use ssh (libssh), yang (libyang) and netconf functionality (libnetconf) directly within the code to implement the Netconf client functionality and attempt to connect to a Netconf server running on M3.

Yes, I believe this is correct. Just to make it clear, because I am not sure how you meant it, implementing basic NETCONF client functionality should be quite simple with *libnetconf2*, you do not need to use *libssh* and maybe not even *libyang* (depends on the RPCs you want to send) directly.

Also, if you have no more questions, please close this issue.

Regards,
Michal
